CI/CD: Github actions

bastimeyer opened the pull request:
## Why

We're currently using two CI providers, Appveyor for running tests on Windows and TravisCI for everything else. While Appveyor has caused issues in the past and is usually much slower than TravisCI, TravisCI's Windows VMs are still in beta with lack of support for secret env vars. See my previous PR #2342.

Github now supports CI/CD workflows via Github actions on Linux, Windows and macOS, which allows us to have all CI tasks being run in one place with direct integration into Github.

Docs can be found here:
https://help.github.com/en/categories/automating-your-workflow-with-github-actions

## Checklist

- [ ] **Re-enable actions**
  I've disabled all actions in the repo settings for now. These changes should be tested in a secondary repo first. See below...
- [x] **`test` job**
  - [x] Run in parallel on ubuntu-latest, macOS-latest and windows-latest, using Python 2.7 and 3.5..3.8
  - [x] Test (pytest)
  - [x] Lint (flake8)
  - [x] Report code coverage (codecov)
    - [x] Add `CODECOV_TOKEN` to secret env vars
- [x] **`documentation` job** (runs in parallel)
- [x] **`windows-installer` job** (runs in parallel)
  - [x] Deploys build artifact on tag with current version
  - [x] Deploys build artifact on scheduled run with current version
- [x] **`deploy-documentation` job**
  Depends on `test`, `documentation` and `windows-installer`.
  Runs when pushed to master or a tag gets pushed. Forks are disabled.
  - [x] Rewrite as bash script with similar/same behavior as doctr
  - [x] Add secret env vars to repo settings
    - [x] `DOCS_DEPLOY_PASSPHRASE`
  - [x] Replace deploy key
    - [x] Add public key to @streamlink-bot
  - [x] Give @streamlink-bot write access to docs repo
  - [x] Add 2FA to @streamlink-bot 
- [x] **`release` job**
  Depends on `deploy-documentation`.
  Runs if tag was pushed. Forks are disabled.
  - [x] Add secret env vars to repo settings
    - [x] `RELEASES_API_KEY`
    - [x] `RELEASE_SIGN_PASSPHRASE`
    - [x] `PYPI_USER`
    - [x] `PYPI_PASSWORD`
  - [x] Replace signing key (same key, different encryption)
  - [x] Deploy to PyPI
    - [x] Split sdist/wheel script into build and deploy script
  - [x] Deploy to Github releases
    - [x] Cache Windows installer artifact
    - [x] Rewrite Github releases helper script with the following features
      - Create or update release
      - Upload assets
      - Read Github's env vars
      - Move release template text into its own file
- [x] **Verify all this in a secondary repo**
  See https://github.com/bastimeyer-test/streamlink-test/actions?query=is%3Asuccess
  - [x] Test job
    Codecov: https://codecov.io/gh/bastimeyer-test/streamlink-test/commit/c10c200520548a27e7fbf08c13709cea23b3b7e7
  - [x] Documentation job
  - [x] Windows installer job
  - [x] Deploy documentation job
    - Latest
      - https://github.com/bastimeyer-test/streamlink-test/runs/557448378
      - https://github.com/bastimeyer-test/streamlink-website-test/commit/d294d345fd6d6c5a6535f82572ef90b49af8f64d
    - Stable
      - https://github.com/bastimeyer-test/streamlink-test/runs/557465088
      - https://github.com/bastimeyer-test/streamlink-website-test/commit/2cdb908a58351d4835288781b1640550191db22c
  - [x] Release job
    - https://github.com/bastimeyer-test/streamlink-test/runs/557468060
    - https://github.com/bastimeyer-test/streamlink-test/releases/tag/2020.4.3
  - [x] Scheduled nightly build
    - https://github.com/bastimeyer-test/streamlink-test/runs/557471458
    - https://github.com/bastimeyer-test/streamlink-test/actions/runs/69723359
- [ ] **Remove old stuff once approved**
  - [ ] TravisCI config
  - [ ] AppVeyor config
  - [x] Old doctr deploy keys
  - [ ] Unused bintray deploy scripts

## Questions

- **How fast is it?**
  Build time changed from >15 mins to <5 mins.
- **Is testing on macOS really needed?**
  This spawns 5 additional jobs, but doesn't slow down the build, as the jobs running on Windows always finish last. Adds to the repo's overall available build time though.
- **Build artifacts**
  The windows installer is now stored as a build artifact, which will be deleted after 90 days. There's an API endpoint available that can be queried for getting the latest artifact. This could be added to the docs, or just a link to the repo's successfully run scheduled actions.

bastimeyer commented:
Tokenless coverage data uploads to Codecov via github actions is now possible.
https://community.codecov.io/t/whitelist-github-action-servers-to-upload-without-a-token/491/17

It's also been a while since I've worked on this PR. I'll see if I can finish this some time soon.

bastimeyer commented:
Apologies for the force-push spamming of unfinished work.

This is still **not finished**, but everything should be in place now. See the updated PR comment and its to-do list. This was a lot of trial-and-error getting this to work, since the Github actions/workflows are not 100% perfect yet and a couple of workarounds were needed. A few very minor things still need to be changed, which I've noticed while updating the comment. Old CI configs and unused build scripts also still need to be removed.

However, feedback and reviews are now welcome. I know it's a lot of changes, especially because most scripts got rewritten. Unfortunately, the diff of a force-push also includes all changes on master (why?), so if I'm going to change something after feedback, this will be drowned in unrelated stuff. I think it'd be better to re-open this PR.

gravyboat commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,165 @@
+        run: git fetch --depth=1 origin +refs/tags/*:refs/tags/*
+      - uses: actions/setup-python@v1
+        with:
+          python-version: 3.5
```
What's the reasoning for building with 3.5 here and below instead of a more recent release?

bastimeyer replied:
The same version is currently set in the travisci config https://github.com/streamlink/streamlink/blob/d4b00527aae0bea34a453b9b0e53caf56d7ac2bc/.travis.yml#L14-L15
I didn't want to introduce more unnecessary changes, but yes, you're right, we should change that.
I'll update this tomorrow in one go with some other things. It'll also be checked/verified on my test repo then.

gravyboat replied:
@bastimeyer It's not a huge deal so don't stress over it, just figured since we're updating so much it's worth a discussion to change it. I'm fine with sticking on 3.5 if you want for a bit.

bastimeyer commented:
Everything is now updated, including the PR comment.

Fixed some issues, updated the python version for build/deploy jobs and tuned some configs/names/messages. Unless there are more issues to fix, that should be it from my side, well, except the missing new docs deploy key, which *should* ideally be included in the last commit.

What needs to be done now is adding the secret env vars to streamlink's repo settings. As I've mentioned in my PR comment, I don't have access to any and can't copy them from travis.

And regarding the new docs deploy key, I can of course generate one myself and send everyone the secret decryption key + iv. The public key also needs to be added to the @streamlink-bot account and I still don't know who's registered that account.

gravyboat commented:
@bastimeyer I've added you, beardypig, and back-to as admins. Not sure why that wasn't already done, but it is now so our bus factor is better. I think @beardypig has the travis keys and the streamlink-bot details? I checked in Travis and while I can see the variables are there, I can't view them, and I don't know what the existing values are.

bastimeyer commented:
> I've added you, beardypig, and back-to as admins to the Streamlink repo.

Since I'm already listed as orga owner, individual repo permissions will be overridden (I think).

> while I can see the variables are there, I can't view them, and I don't know what the existing values are.

That's the default behavior of most (or all) CI providers, so nobody can steal the data if they manage to steal login data of authorized people. That's why this data has to be stored locally (encrypted).

Only a different version of the signing key was allegedly shared between some of you in #1701. If @beardypig doesn't have the key+iv of the encrypted passwordless signing key anymore, then he'll need to encrypt it again with new parameters and add these to the repo settings, as described above.

**Just to sum up all missing secure env vars again**

- `CODECOV_TOKEN`
  Codecov lets anyone with access copy this from their site. Will add this now.
- `DOCS_DEPLOY_KEY` / `DOCS_DEPLOY_IV`
  As mentioned earlier, I can generate a new deploy key myself and send it to everyone, assuming the data will be kept safe. The public key of the decrypted RSA key needs to be added to @streamlink-bot.
- `RELEASES_API_KEY`
  @streamlink-bot simply needs to generate a new OAuth token (with the repo scope) from the account settings.
- `RELEASE_SIGN_KEY` / `RELEASE_SIGN_IV`
  The key decryption data we've just been talking about. The env vars were named `encrypted_eeb8b970d3a3_{key,iv}` on TravisCI and I renamed them for better understanding.
- `PYPI_USER`
  Value is `streamlink`. Will add this now.
- `PYPI_PASSWORD`
  I don't have it. Listed as maintainers of the PyPI package are @cdrage, @beardypig and @gravyboat.

gravyboat commented:
@bastimeyer Yeah I'm aware they are encrypted by default, I'm just noting it so no one asks. I double checked my encrypted documents and I don't see the signing keys anywhere unfortunately. I don't have access to the @streamlink-bot account as far as I know. I do have the pypi_password and have added it to the repo secrets.

bastimeyer commented:
@beardypig is the owner of @streamlink-bot 
https://github.com/streamlink/streamlink/issues/899#issuecomment-400373602

Looks like he's currently having a break or so, which is absolutely fine. Should he not reply here within the next week or so, I'll send him an email and ask for the necessary data.

And in the case he shouldn't reply to that either, we \*could\* run a build on TravisCI from another branch, so that secret env vars are included, store them in a file, encrypt that file with a new key, upload it somewhere and decrypt it locally. That way we don't leak anything and don't have to replace the encrypted signing key, which could potentially cause confusion to some end-users browsing the git log. The only problem left would be @streamlink-bot, unless we use a different/new account.

beardypig commented:
@bastimeyer @gravyboat checking in

beardypig commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,165 @@
+          name: ${{ steps.installer.outputs.filename }}-win32
+          path: ${{ env.STREAMLINK_DIST_DIR }}
+      - name: sdist and wheels
+        env:
+          RELEASE_SIGN_KEY: ${{ secrets.RELEASE_SIGN_KEY }}
+          RELEASE_SIGN_IV: ${{ secrets.RELEASE_SIGN_IV }}
```
```suggestion
        env:
          RELEASE_KEY_PASSPHRASE: ${{ secrets. RELEASE_KEY_PASSPHRASE }}
```

beardypig commented on script/build-and-sign.sh:
```diff
@@ -0,0 +1,61 @@
+
+if [[ "${CI}" = true ]] || [[ -n "${GITHUB_ACTIONS}" ]]; then
+    echo "build: Decrypting signing key" >&2
+    openssl aes-256-cbc \
+        -K "${RELEASE_SIGN_KEY}" \
+        -iv "${RELEASE_SIGN_IV}" \
+        -in "${KEY_FILE_ENC}" \
+        -out "${KEY_FILE}" \
+        -d
```
I have changed the way the file is encrypted, using the [suggested method by GitHub](https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets#limits-for-secrets). The encrypted file will need to be updated in the repo too. I do not have the original IV and key for the encrypted file as I seem to remember it was done automatically by the travis CLI and I don't think the IV/key were ever exposed to me either. 

```suggestion
	gpg --quiet --batch --yes --decrypt \
		--passphrase="${RELEASE_KEY_PASSPHRASE}" \
		--output "${KEY_FILE}" \
		"${KEY_FILE_ENC}"
```

beardypig commented:
I figured the best way to contribute changes to this was to open a PR on your branch @bastimeyer, as I needed to include new files etc. 

We'll have to work out a secure way to share the signing key and credentials for the @streamlink-bot account.

bastimeyer commented:
Glad to hear from you back, @beardypig.

No problem, we can change the decryption stuff. This is fine.

Could you please add/push your encrypted signing key to the PR branch and also create a new passwordless RSA 4096 key, encrypt it the same way (with a different passphrase) and add it to the branch as well? That should be sufficient. I can pull and rebase the keys then to the proper commits.

All you have to do afterwards then is adding the new passphrases to the secret env var list on the streamlink repo here. And also don't forget the @streamlink-bot OAuth token, so that the account can create releases here.

bastimeyer commented:
Ok, didn't see your PR on my fork. This is fine too.
Will check this later.

beardypig commented:
Can I put directly to the branch? I've never worked out how to do that :|

bastimeyer commented:
Keys have been updated now and all the necessary secret env vars have been added to the repo. @beardypig also helped setting the env vars on my test repo (thanks github for making it unnecessarily difficult with manual API requests), so we could verify that everything is working.

I've updated my initial PR comment accordingly with the new workflow and test-release / docs deploy links.

Apparently deleting the other CI configs isn't something I can do here, so I will fix that now.

**What's left to do**

- [ ] remove old CI configs
- [ ] give streamlink-bot write permissions to the docs repo
- [ ] 2FA on the streamlink-bot account
- [ ] Update docs with new links to nightly builds. This should be done in two steps ideally, since we haven't built any nightly build artifacts yet obviously, and there's no build history because of that.
- [ ] finally re-enable actions on the repo and merge this

bastimeyer commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,164 @@
+      - name: Github release
+        env:
+          RELEASES_API_KEY: ${{ secrets.RELEASES_API_KEY }}
+        run: ./script/github_releases.py "${STREAMLINK_DIST_DIR}/"*".exe"
```
Just noticed that the sdist file and signature don't get uploaded to the github release:
https://github.com/streamlink/streamlink/blame/master/.travis.yml#L96

Is this something we still want to do, or is it enough to publish this on PyPI?
```bash
curl -s 'https://api.github.com/repos/streamlink/streamlink/releases/latest' | jq '[.assets[] | {name:.name, count:.download_count}]'
```
```json
[
  {
    "name": "streamlink-1.3.1.exe",
    "count": 23947
  },
  {
    "name": "streamlink-1.3.1.tar.gz",
    "count": 1019
  },
  {
    "name": "streamlink-1.3.1.tar.gz.asc",
    "count": 158
  }
]
```

back-to replied:
Arch Linux as example uses a github URL
https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/streamlink#n15

so it might be useful

bastimeyer replied:
Already fixed it, just rebuilding the test release before updating the PR to be extra sure.

bastimeyer commented:
Also, I've asked about this above, do we really need to run tests on macOS (all python versions)? There's nothing specific to be tested on macOS AFAIK and it's just POSIX vs win32 if I'm not mistaken. Running 5 additional jobs can take additional time, it streches the list of merge requirements in PRs a bit and uses around 1/3 of the available monthly build time should we ever get close to that limit (probably not, but not sure either in case of abuse).

gravyboat commented:
@bastimeyer Even if it's nearly the same I think it's still worth running rather than potentially running in to an issue later because of some stupidity that Apple has decided as they seem to be increasing their rate of bad decisions. If it becomes a problem we can cut it, but I'd rather it just run than have to do some sort of emergency patch after a release because of a mac problem.

beardypig commented:
I would drop macOS from the matrix, there might be some differences in paths but that is tested anyway.

bastimeyer commented:
We've not been running tests on macOS with the current CI setup the whole time, so I don't think it's necessary to do it now. I also think that Github doesn't allocate the available CI runner instances evenly, because the macOS runners were being queued yesterday whereas the ubuntu and windows ones were not. Also, the macOS runners have become very slow and always finish last, which wasn't the case when I first ran the workflows. I'll keep it for now though, and if it gets annoying, we can remove macOS in another PR.

----

So, now that Github is working again, I've just built another test-release and also run another scheduled test build (the one from yesterday didn't work because of the wrong repo name).

Initial PR comment is updated with the new links.

bastimeyer commented:
Unless somebody has found another issue or has other/further concerns, we should be good to go now.

Before this gets merged, actions have to be re-enabled in the repo-settings.

Once we've merged this, TravisCI and AppVeyor webhooks/integrations (and remaining bintray configs) can be removed, so we can delete the configs afterwards without making it fail.

All currently open PRs will then need to be rebased to master (or master needs to be merged in). I've added the "needs rebase" label for this.

I will also send another PR later today with the updated docs regarding the windows installer nightly build downloads.

And that should be it. CI migration done. :tada: :tada: :tada:

beardypig commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,164 @@
+        run: bash ./script/install-dependencies.sh
+      - name: Test
+        continue-on-error: ${{ matrix.continue }}
+        run: pytest --cov
```
Can we include the `-ra` option for `pytest` to output some extra summary info at the end of the tests? It shows some extra details about why tests were skipped or why failures were allowed. 

```suggestion
        run: pytest -ra --cov
```

bastimeyer replied:
Done. Added this as another commit, as it's not been set previously.

gravyboat commented:
Looks great! Merging, thanks @bastimeyer.

bastimeyer commented:
I meant that all other PRs need to be rebased now. Well, after the other CIs have been removed. As you can see, TravisCI is failing now because of the doctr removal.

I will now remove the TravisCI/AppVeyor webhooks/integrations.

bastimeyer commented:
I've only deleted the webhooks for now. I think removing the Github App integration (which is the old and deprecated way of connecting it) will delete its build history. I've also disabled the triggers and cron on the TravisCI account settings. I think this should be fine. Will send a PR with the removal of the old stuff now.

back-to commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,164 @@
+        run: ./script/github_releases.py "${STREAMLINK_DIST_DIR}/"*{.exe,.tar.gz{,.asc}}
+      - name: PyPI release
+        env:
+          PYPI_USER: ${{ secrets.PYPI_USER }}
```
the build log is just `***` for streamlink
https://github.com/streamlink/streamlink/runs/602730028?check_suite_focus=true#step:5:12
https://github.com/streamlink/streamlink/runs/602729990?check_suite_focus=true#step:7:1

I don't think a secret is needed here

bastimeyer replied:
True. I was already wondering where this was coming from. Can you send a PR with the fix, please?
